conf_file.h

```diff
@@ -17,24 +17,123 @@
+                        void *data);
+    /**
+     * Loop to all sub-sections from the section @a section_name
+     *
+     * @param section_name The name of the section to look for sub-sections
+     * @param separator Charatecter that separates a section from the sub-section
+     * @param sub_name A pointer to a char* that will be updated with current sub-section name
+     * @param sub_len A pointer to size_t element be filled with the length of the current
+     * sub-section name.
+     * @param table Array of TableOption with information on which fields are going to be
+     * extracted and how to extract them.
+     * @param table_len The number of elements in @a table.
+     * @param data A pointer to the struct that will be used to hold the extracted data.
+     * @param ptr A void pointer to be used internally in the loop. Must be initialized with @c NULL
+     * for first execution.
+     */
+    int loop_sub_sections(const char *section_name, char separator, const char **sub_name,
+                          size_t *sub_len, const OptionsTable table[], size_t table_len, void *data,
+                          void **ptr);
+
+    // Helpers
+    static int parse_bool(const char *val, size_t val_len, void *storage, size_t storage_len);
+    static int parse_str_dup(const char *val, size_t val_len, void *storage, size_t storage_len);
+    static int parse_str_buf(const char *val, size_t val_len, void *storage, size_t storage_len);
+    DECLARE_PARSE_INT(i);
+    DECLARE_PARSE_INT(ul);
+    DECLARE_PARSE_INT(ull);
 
 private:
-    const char *_filename;
-    struct section *_sections = nullptr;
-    struct section *_current_section = nullptr;
+    struct conffile *_files;
+    struct section *_sections;
+
+    int _parse_file(const char *addr, size_t len, const char *filename);
+    struct section *_find_section(const char *section_name, size_t len);
+    struct config *_find_config(struct section *s, const char *key_name, size_t key_len);
 
-    int _add_section(char *entry, int line);
-    int _add_config(char *entry, int line);
-    void _trim(char *str);
+    struct section *_add_section(const char *addr, size_t len, int line, const char *filename);
+    int _add_config(struct section *s, const char *entry, size_t entry_len, const char *filename,
+                    int line);
+    void _trim(const char **str, size_t *len);
+    int _extract_options_from_section(struct section *s, const OptionsTable table[],
+                                      size_t table_len, void *data);
 };
+
+/*
+ * Return the number of elements in a OptionTable array.
+ */
+#define OPTIONS_TABLE_SIZE(_table) (sizeof(_table) / sizeof(ConfFile::OptionsTable))
+
+/*
+ * Create the last two elements in OptionsTable, storage_offset and storage_len, using _field
+ * from _struct.
+ */
+#define OPTIONS_TABLE_STRUCT_FIELD(_struct, _field) \
+    offsetof(_struct, _field), sizeof(_struct::_field)
```

Comment:
I think it's better to define OptionsTable with sub-struct containing offset and size. Then this would become:
```c++
#define OPTIONS_TABLE_STRUCT_FIELD(_struct, _field) \
     { offsetof(_struct, _field), sizeof(_struct::_field) }
```

This way it's easier to check that the number of fields initialized in the struct match the number of fields it should be.